ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFTrunc.java

```diff
@@ -477,21 +477,38 @@ protected DoubleWritable trunc(DoubleWritable input, int scale) {
   }
 
   protected BigDecimal trunc(BigDecimal input, int scale) {
```

Comment:
The calculations below are somewhat complex and resemble a bit what is happening inside `BigDecimal#setScale`. Can we replace some of the logic below with calls to `BigDecimal#setScale`?